File: packages/sandbox/daemon/routes/bash.ts
import type { TaskManager } from "../process/task-manager";
import { jsonResponse, parseBase64JsonBody } from "./body-parser";
import { awaitTaskResponse } from "./tasks";

const DEFAULT_TIMEOUT_MS = 30_000;
const MAX_TIMEOUT_MS = 15 * 60 * 1000;

export type BashMode = "await" | "background";

export interface BashDeps {
  /** Default cwd for unscoped commands. Typically `<appRoot>/repo` (the repo). */
  repoDir: string;
  taskManager: TaskManager;
  env?: Record<string, string>;
}

interface BashBody {
  command?: string;
  timeout?: number;
  cwd?: string;
  env?: Record<string, string>;
  mode?: BashMode;
}

/**
 * Modes:
 *   - "await" (default): runs to completion and returns the full
 *     stdout/stderr/exitCode body. This is the legacy bash behavior.
 *   - "background": returns the taskId immediately. Caller can poll
 *     /_decopilot_vm/tasks/:id, stream output, or kill via the tasks API.
 */
export function makeBashHandler(deps: BashDeps) {
  return async (req: Request): Promise<Response> => {
    let body: BashBody;
    try {
      body = (await parseBase64JsonBody(req)) as BashBody;
    } catch (e) {
      return jsonResponse({ error: (e as Error).message }, 400);
    }
    if (!body.command || typeof body.command !== "string") {
      return jsonResponse({ error: "command is required" }, 400);
    }

    const mode: BashMode = body.mode === "background" ? "background" : "await";
    const timeout = clampTimeout(body.timeout, mode);
    const env = body.env ? { ...(deps.env ?? {}), ...body.env } : deps.env;

    const task = await deps.taskManager.spawn({
      command: body.command,
      cwd: body.cwd ?? deps.repoDir,
      env,
      mode: "pipe",
      timeoutMs: timeout,
      label: `$ ${body.command}`,
    });

    if (mode === "background") {
      return jsonResponse({ taskId: task.id, status: task.status });
    }

    return awaitTaskResponse(deps.taskManager, task.id, {
      timedOutExitCode: -1,
    });
  };
}

function clampTimeout(raw: number | undefined, mode: BashMode): number {
  const fallback = DEFAULT_TIMEOUT_MS;
  const requested = typeof raw === "number" && raw > 0 ? raw : fallback;
  // Background tasks may run longer; cap at 15 min (matches TaskManager TTL).
  const ceiling = mode === "background" ? MAX_TIMEOUT_MS : 120_000;
  return Math.min(requested, ceiling);
}

File: packages/sandbox/daemon-go/internal/routes/bash.go
package routes

import (
	"net/http"

	"github.com/decocms/studio/sandbox-daemon/internal/httpx"
	"github.com/decocms/studio/sandbox-daemon/internal/proc"
)

const (
	bashDefaultTimeoutMs    = 30_000
	bashAwaitCeilingMs      = 120_000
	bashBackgroundCeilingMs = 15 * 60 * 1000
)

type BashDeps struct {
	RepoDir     string
	TaskManager *proc.TaskManager
}

func awaitTaskResponse(w http.ResponseWriter, tm *proc.TaskManager, id string, extra map[string]any, timedOutExitCode *int) {
	result, ok := tm.Finished(id)
	if !ok {
		httpx.Error(w, 500, "task vanished before completion")
		return
	}
	out, _ := tm.Output(id)
	exitCode := result.ExitCode
	if timedOutExitCode != nil && result.TimedOut {
		exitCode = *timedOutExitCode
	}
	resp := map[string]any{
		"stdout":    out.Stdout,
		"stderr":    out.Stderr,
		"exitCode":  exitCode,
		"timedOut":  result.TimedOut,
		"truncated": out.Truncated,
	}
	for k, v := range extra {
		resp[k] = v
	}
	httpx.JSON(w, 200, resp)
}

func Bash(deps BashDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var body struct {
			Command string            `json:"command"`
			Timeout int               `json:"timeout"`
			Cwd     string            `json:"cwd"`
			Env     map[string]string `json:"env"`
			Mode    string            `json:"mode"`
		}
		if err := decodeBody(r, &body); err != nil {
			httpx.Error(w, 400, err.Error())
			return
		}
		if body.Command == "" {
			httpx.Error(w, 400, "command is required")
			return
		}
		mode := "await"
		if body.Mode == "background" {
			mode = "background"
		}
		timeout := bashDefaultTimeoutMs
		if body.Timeout > 0 {
			timeout = body.Timeout
		}
		ceiling := bashAwaitCeilingMs
		if mode == "background" {
			ceiling = bashBackgroundCeilingMs
		}
		if timeout > ceiling {
			timeout = ceiling
		}
		cwd := deps.RepoDir
		if body.Cwd != "" {
			cwd = body.Cwd
		}

		task := deps.TaskManager.Spawn(proc.TaskSpec{
			Command:   body.Command,
			Cwd:       cwd,
			Env:       body.Env,
			Mode:      "pipe",
			TimeoutMs: timeout,
			Label:     "$ " + body.Command,
		})

		if mode == "background" {
			httpx.JSON(w, 200, map[string]any{"taskId": task.ID, "status": task.Status})
			return
		}
		timedOut := -1
		awaitTaskResponse(w, deps.TaskManager, task.ID, nil, &timedOut)
	}
}
